### PR TITLE
UDP client/server

### PR DESCRIPTION
**Pull request by mfine**
Similar to TCP client/server, but in UDP.

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-develop-2017082920-3-gaa6099b
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017082920-3-gaa6099b/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-develop-2017082920-3-gaa6099b/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017082920-3-gaa6099b

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-develop-2017082920-4-g3b1e790
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017082920-4-g3b1e790/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-develop-2017082920-4-g3b1e790/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017082920-4-g3b1e790

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-develop-2017102720-2-ga42b9db
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017102720-2-ga42b9db/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-develop-2017102720-2-ga42b9db/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017102720-2-ga42b9db

**Comment by mfine** [comment]
This is working now - ready for a review.

/cc @lbara @silverjam

File: package/zmq_adapter/src/zmq_adapter_udp_connect.c [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+  while (1) {
+    int fd = socket_create(&s_addr);
+    if (fd < 0) {
+      debug_printf("error connecting UDP socket\n");
+      sleep(CONNECT_RETRY_TIME_s);
+      continue;
+    }
+
+    io_loop_start(-1, fd);
```
Comment (mfine) [inline package/zmq_adapter/src/zmq_adapter_udp_connect.c]: @denniszollo @lbara @silverjam I think we should use this similarly in the TCP client to only receive data for the piksi and NOT send the piksi data out.

Comment (mfine) [reply package/zmq_adapter/src/zmq_adapter_udp_connect.c]: Was going to say the same thing about the TCP server, but that's a bi-directional link :(

**Comment by mfine** [comment]
The changes here add a UDP client/server functionality somewhat similar to the TCP client/server. Differences are that these are not bi-directional communication - they either only publish or subscribe to things, not both. The weirdness of the TCP client also being bi-directional might be worth fixing at the same time.

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-develop-2017102720-2-gf424b14
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017102720-2-gf424b14/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-develop-2017102720-2-gf424b14/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-develop-2017102720-2-gf424b14

**Comment by silverjam** [comment]
@mfine to close down tcp bi directionality we need to see if we can call shutdown() on the socket with SHUT_WR

**Comment by mfine** [comment]
@silverjam I'm having second thoughts on removing bi-directionality from TCP!!! Mainly, because it seems odd to line it up with UDP behavior - that would mean that the TCP client pushes its content to a server (not receives content). Going to leave things as is for now around that behavior.

File: package/zmq_adapter/src/zmq_adapter_udp_connect.c [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright (C) 2016 Swift Navigation Inc.
```
Comment (silverjam) [inline package/zmq_adapter/src/zmq_adapter_udp_connect.c]: Update header?

**Pull request merged**